lib/charms/operator_libs_linux/v0/apt.py

```diff
@@ -1314,6 +1318,203 @@ def _parse(line: str, filename: str) -> DebianRepository:
         else:
             raise InvalidSourceError("An invalid sources line was found in %s!", filename)
 
+    def load_deb822(self, filename: str) -> None:
+        """Load a deb822 format repository source file into the cache.
+
+        Args:
+          filename: the path to the repository file
+
+        In contrast to one-line-style, the deb822 format specifies a repository
+        using a multi-line paragraph. Paragraphs are separated by whitespace,
+        and each definition consists of lines that are either key: value pairs,
+        or continuations of the previous value.
+
+        Read more about the deb822 format here:
+            https://manpages.ubuntu.com/manpages/noble/en/man5/sources.list.5.html
+        For instance, ubuntu 24.04 (noble) lists its sources using deb822 style in:
+            /etc/apt/sources.list.d/ubuntu.sources
+
+        The semantics of `load_deb822` slightly different to `load`:
+            `load` calls `_parse`, which reads a commented out line as an entry that is not enabled
+            `load_deb822` strips out comments entirely when parsing a file into paragraphs, and
+                assumes that comments have been removed when parsing individual paragraphs/entry,
+                instead only reading the 'Enabled' key to determine if an entry is enabled
+        """
+        with open(filename, "r") as f:
+            repos, errors = self._parse_deb822_lines(f, filename=filename)
+
+        for repo in repos:
+            repo_identifier = "{}-{}-{}".format(repo.repotype, repo.uri, repo.release)
+            self._repository_map[repo_identifier] = repo
```

Comment:
I wonder if there should be a new class - e.g. something like `Debian822Repository` - or a new attribute on `DebianRepository` - e.g. `DebianRepository(...).source_format` - for sources that are written in deb822 style format :thinking: 

Reason saying is that I think the current `DebianRepository` class is well suited for one-line style sources, but not deb822 style. If a particular `DebianRepository` generated from the deb822 source, how does that get mapped back correctly into the `*.sources` file? Even though we're able to convert the deb822 style source to multiple one-lines, the `RepositoryMapping` cache loses how the source is actually represented in the `*.sources` file it was read from. Perhaps if we had some tag or differentiation between the two formats on `DebianRepository`, we can signal that the source should be treated different when dumping a source into a file under `/etc/apt/sources.list.d/` or modifying an existing source in a `*.source` file.

Reply:
I think this is actually a good case for inheritance for maximum backwards compatibility. A `Debian822Repository` is a `DebianRepository`, just with different implementation details of some methods.

This is perhaps a good point to think about how to handle a `Signed-by` key that was provided inline. It's not clear to me how best to handle the existing `gpg_key` property in this case, which is expected to be a string filename. There doesn't seem to be a property for getting the key file contents directly -- the other key related methods are all about adding a key to a repository.

 Here are some options for how we could handle the inline key case.
1. return an empty string (default value), and document that users should check `gpg_key_inline` in this case
    - old code won't instantly break
    - old code will silently assume that inline key entries aren't signed
        - maybe ok if they raise an error in this case, since they can then be fixed (but bad to break them)
        - bad if instead you get a silent behaviour change (seems more likely to me, but idk)
    - new code will have to do `isinstance` or `hasattr` or catch `AttributeError` when checking for `gpg_key_inline`
3. return a new sentinel value to indicate that you should check `gpg_key_inline`
    - instantly breaks old code expecting a string
    - doesn't hide any errors with treating the entry as unsigned
    - new code has to switch based on the return value (but probably is already to handle the empty string case indicating no key)
5. just return the inline key instead of a filename
    - instantly breaks old code expecting a filename
    - doesn't hide any errors with treating the entry as unsigned
    - annoying for all users to check if it's a filename or key
7. try to do something clever by writing the inline key to a temporary file (in memory?) and returning the path to that
    - old code works perfectly
    - added complexity? what could go wrong here?